proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,262 @@
+to the entry function.
+
+The default mode does not issue diagnostics for `CallExpr` nodes that are inside
+functions which are not reachable from exported functions.
+
+### Relaxed Diagnostic Mode
+
+The implementation of the relaxed diagnostic mode matches the default mode,
+except that when a `CallExpr` references an unavailable API, the compiler emits
+a _warning_.
+
+The relaxed mode does not issue diagnostics for `CallExpr` nodes that are inside
+functions which are not reachable from exported functions. A user enables
+relaxed mode by passing `-Wno-error=hlsl-availability`.
+
+### Strict Diagnostic Mode
+
+The strict diagnostic mode strives to aggressively issue diagnostics. For
+non-library shaders, during parsing any callee of a `CallExpr` that has
+availability annotation that marks it as unavailable for the target shader model
+and stage will produce an _error_.
+
+For library shaders, during parsing any callee of a `CallExpr` that has
+availability annotation that marks it as unavailable for the target shader model
+version will produce an _error_. After parsing the translation unit, an AST
+visitor will traverse to all `CallExpr` nodes that are reachable from annotated
+shader entry functions. If the callee of a `CallExpr` has availability
+annotations that signify that the API is unavailable for the target shader stage
+the compiler emits an _error_.
+
+Unlike in the _default_ or _relaxed_ mode, the compiler will emit diagnostics
+for mismatched shader model version without the use of a call graph and
+regardless of reachability.
+
+A user enables strict mode by passing `-fhlsl-strict-diagnostics`. 
+
+### Comparison Against Existing Behavior
+
+Today DXC allows the use of APIs in shaders regardless of the specified target
+profile. The responsibility for enforcing API availability falls to the bytecode
+validator. This results in compilation passing but the validator failing. For
+example take the following code:
+
+> [Godbolt Link](https://godbolt.org/z/v1sjEEETW)
+```c++
+Texture2D texture : register(t0);
+SamplerState samplerState : register(s0);
+
+FeedbackTexture2D<SAMPLER_FEEDBACK_MIN_MIP> map : register(u0);
+
+[numthreads(4, 4, 1)] 
+void main(uint3 threadId : SV_DispatchThreadId) {
+  float2 uv = threadId.xy;
+  uv /= 256;
+
+  map.WriteSamplerFeedbackLevel(texture, samplerState, uv, threadId.x % 8);
+}
+```
+
+Compiled with the `-T cs_6_4` flag this produces the validation error:
+```
+<>:11:3: error: Opcode WriteSamplerFeedbackLevel not valid in shader model cs_6_4.
+note: at 'call void @dx.op.writeSamplerFeedbackLevel(i32 176, %dx.types.Handle %1, %dx.types.Handle %2, %dx.types.Handle %3, float %8, float %9, float undef, float undef, float %11)' in block '#0' of function 'main'.
+Validation failed.
+```
+
+This error displays the LLVM IR text for the error site. In a trivial example
+like the one above finding the error is easy. In a complex shader that could be
+thousands of lines long this is unmanageable.

```

Comment:
Note that the error does have the source location, but since way the compiler is used here provides no source file name, you only get `<>:11:3:`.

Reply:
Yes, that's true for all the examples not just the DXC validator one. Godbolt gives the filename as `<source>`, I truncated that down to just `<>` for brevity since it isn't relevant.

While I think that generally this class of diagnostics (availability of DXIL operations) is one of the more-stable diagnostic locations for the DXIL validator, the diagnostic would be way more approachable to users if it was generated on the AST and referred to the HLSL source. That's the core goal of this proposal.